src/main/java/com/eleks/academy/whoami/core/state/SuggestingCharacters.java

```diff
@@ -90,65 +89,43 @@ public Map<String, SynchronousPlayer> getPlayers() {
 		return this.players;
 	}
 
-	/**
-	 * The term author is referred to a player who suggested at least one character
-	 * <p>
-	 * Basic algorithm description:
-	 * 1) Collect all randomly-ordered authors into a form of cyclic oriented graph.
-	 * 2) Assign to each author a random character suggested by the next author (next graph node)
-	 * 3) Randomly assign all the suggested characters that are left disregarding the author to
-	 * all the non-author players
-	 */
-	private GameState assignCharacters() {
-		Function<String, Integer> randomAuthorOrderComparator = value ->
-				Double.valueOf(Math.random() * 1000).intValue();
-
-		final var authors =
-				this.suggestedCharacters.keySet()
-						.stream()
-						.sorted(Comparator.comparing(randomAuthorOrderComparator))
-						.collect(Collectors.toList());
-
-		authors.forEach(author -> {
-			final var character = this.getRandomCharacter()
-					.apply(this.suggestedCharacters.get(this.<String>cyclicNext().apply(authors, author)));
-
-			character.markTaken();
-
-			this.playerCharacterMap.put(author, character.getCharacter());
-		});
+	Map<String, String> assignCharacters() {
+		Map<String, String> playerToCharacterCopy = new HashMap<>(playerCharacterMap);
 
-		final var authorsSet = new HashSet<>(authors);
+		boolean isValueChanged;
 
-		final var nonTakenCharacters = this.suggestedCharacters.values()
-				.stream()
-				.flatMap(Collection::stream)
-				.filter(character -> !character.isTaken())
-				.collect(toList());
+		do {
+			playerToCharacterCopy = getRandomCharacter(playerToCharacterCopy);
+			isValueChanged = isTwoValueEquals(playerToCharacterCopy);
+		} while (!isValueChanged);
 
-		this.players.keySet()
-				.stream()
-				.filter(player -> !authorsSet.contains(player))
-				.forEach(player -> {
-					final var character = this.getRandomCharacter().apply(nonTakenCharacters);
+		return playerToCharacterCopy;
+	}
 
-					character.markTaken();
+	private Map<String, String> getRandomCharacter(Map<String, String> playerCharacter) {
+		List<String> key = new ArrayList<>(playerCharacter.keySet());
+		List<String> value = new ArrayList<>(playerCharacter.values());
 
-					this.playerCharacterMap.put(player, character.getCharacter());
+		Collections.shuffle(value);
 
-					nonTakenCharacters.remove(character);
-				});
+		playerCharacter = IntStream.range(0, playerCharacter.size()).boxed()
+				.collect(Collectors.toMap(key::get, value::get));
 
-		return this;
+		return playerCharacter;
 	}
 
+	private boolean isTwoValueEquals(Map<String, String> playerCharacterShuffled) {
```

Comment:
please refactor this method to except two params old and shuffled, to compare them, such a case would work if it is a public method and it compares the value of current to input, but your method is not so, and no need to do it like this. So the best is to have two parameters and work with them

Reply:
Resolved